### PR TITLE
EASY-1889 return a 415 when a wrong Content-Type is given.

### PR DESCRIPTION
**Pull request by stephanHogenboom**
- [x] case sensitivity
- [x] refine code

Fixes EASY-1889

#### When applied it will...
* return a 415 when a bag is put with another Content-Type than 'application/zip'
* 
#### Related pull requests on github

repo                       | PR
-------------------------- | -----------------
easy-                      | [PR#](PRlink)

### CODE REVIEW

File: src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala [hunk 1]
```diff
@@ -147,21 +147,28 @@ trait StoresServletComponent extends DebugEnhancedLogging {
       trace(())
       basicAuth()
       debug("Authenticated")
-      val bagstore = params("bagstore")
+      val bagStore = params("bagstore")
       val uuidStr = params("uuid")
-      logger.info(s"Received PUT request for bagstore $bagstore and UUID $uuidStr")
-      bagStores.getBaseDirByShortname(bagstore)
+      val contentType = Option(request.getHeader("Content-Type"))
+      logger.info(s"Received PUT request for bagstore $bagStore and UUID $uuidStr")
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala]: @jo-pol is this logging still necessary? I thought we had servlet logging here, right?

File: src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala [hunk 2]
```diff
@@ -147,21 +147,28 @@ trait StoresServletComponent extends DebugEnhancedLogging {
       trace(())
       basicAuth()
       debug("Authenticated")
-      val bagstore = params("bagstore")
+      val bagStore = params("bagstore")
       val uuidStr = params("uuid")
-      logger.info(s"Received PUT request for bagstore $bagstore and UUID $uuidStr")
-      bagStores.getBaseDirByShortname(bagstore)
+      val contentType = Option(request.getHeader("Content-Type"))
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala]: can you rename `contentType` to `requestContentType`? the former is a method name in Scalatra...

File: src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala [hunk 3]
```diff
@@ -147,21 +147,28 @@ trait StoresServletComponent extends DebugEnhancedLogging {
...
+      logger.info(s"Received PUT request for bagstore $bagStore and UUID $uuidStr")
+      bagStores.getBaseDirByShortname(bagStore)
         .map(base => {
           Try { UUID.fromString(uuidStr) }
             .recoverWith {
               case _: IllegalArgumentException => Failure(new IllegalArgumentException(s"invalid UUID string: $uuidStr"))
             }
+            .flatMap(uuid => //TODO make a method in
+              if (contentType.fold(true)(!_.equalsIgnoreCase("application/zip")))
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala]: I'm not completely certain, but this if-else construction sounds a lot like:

```scala
contentType.withFilter(_.equalsIgnoreCase("application/zip"))
    .map(_ => uuid)
    .getOrElse(Failure(UnsupportedMediaTypeException(contentType.getOrElse("none"), "application/zip")))
```

Please verify with tests though 😉

File: src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala [hunk 4]
```diff
@@ -353,10 +353,15 @@ class StoresServletSpec extends TestSupportFixture
     List("Authorization" -> s"$authType $encoded")
   }
 
+  def authenticationHeaderAndContentTypeZip(username: String, password: String): List[(String, String)] = {
+    authenticationHeader(username, password) ::: List(("Content-Type", "application/zip"))
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala]:
```suggestion
    authenticationHeader(username, password) :+ "Content-Type" -> "application/zip"
```

File: src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala [hunk 5]
```diff
@@ -353,10 +353,15 @@ class StoresServletSpec extends TestSupportFixture
...
+  }
+
   private val basicAuthentication = authenticationHeader(username, password)
+  private val basicAuthenticationAndZipContentType = authenticationHeaderAndContentTypeZip(username, password)
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala]: just curious, why not inline that method call?

```scala
private val basicAuthenticationAndZipContentType = authenticationHeader(username, password) :+ "Content-Type" -> "application/zip"
```

**Review by rvanheest — CHANGES_REQUESTED** [review]
nice start!
Question: shouldn't you also add this `415` to the API docs?

File: src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala (hunk 3 above)
Comment (stephanHogenboom) [reply src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala]: `.flatMap(uuid =>
              requestContentType.withFilter(_.equalsIgnoreCase("application/zip"))
                .map(_ => Success(uuid))
                .getOrElse(Failure(UnsupportedMediaTypeException(requestContentType.getOrElse("none"), "application/zip")))
            )`
This compiles and works

File: src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala (hunk 5 above)
Comment (stephanHogenboom) [reply src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala]: Simple, forgot to do that

File: src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala (hunk 3 above)
Comment (stephanHogenboom) [reply src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala]: The code looks much nicer now 👍

Comment (rvanheest) [reply src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala]: extract method?

File: src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala (hunk 1 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala]: It looks like you are right, we don't need it any more.
image

Comment (stephanHogenboom) [reply src/main/scala/nl.knaw.dans.easy.bagstore/server/StoresServletComponent.scala]: Than ill will remove the log line

File: src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala [hunk 6]
```diff
@@ -569,9 +569,25 @@ class StoresServletSpec extends TestSupportFixture
...
+    val uuid = "11111111-1111-1111-1111-111111111111"
+    put(s"/store1/bags/$uuid", body = Files.readAllBytes(testBagUnprunedA), basicAuthentication) {
+      status shouldBe 415
+      body shouldBe "media type none is not supported by this api. Supported types are 'application/zip'"
+    }
+  }
+
+  it should "fail, returning a 415 when an invalid content-type header is given" in {
+    val uuid = "11111111-1111-1111-1111-111111111111"
+    put(s"/store1/bags/$uuid", body = Files.readAllBytes(testBagUnprunedA), basicAuthentication ::: List(("Content-Type", "application/tar"))) {
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala]:
```suggestion
    put(s"/store1/bags/$uuid",
      body = Files.readAllBytes(testBagUnprunedA),
      headers = "Content-Type" -> "application/tar" :: basicAuthentication) {
```

File: src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala [hunk 7]
```diff
@@ -366,9 +366,10 @@ class StoresServletSpec extends TestSupportFixture
   }
 
   private val basicAuthentication = authenticationHeader(username, password)
+  private val basicAuthenticationAndZipContentType = basicAuthentication  :+ "Content-Type" -> "application/zip"
 
   def putBag(uuid: String, bagZip: Path): Unit = {
-    put(s"/store1/bags/$uuid", body = Files.readAllBytes(bagZip), basicAuthentication) {
+    put(s"/store1/bags/$uuid", body = Files.readAllBytes(bagZip), basicAuthenticationAndZipContentType) {
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala]: could you add `headers = ` to every one of these? just for clearance in reading.

```scala
put(s"/store1/bags/$uuid", body = Files.readAllBytes(bagZip), headers = basicAuthenticationAndZipContentType) {
```

File: src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala [hunk 8]
```diff
@@ -366,9 +366,10 @@ class StoresServletSpec extends TestSupportFixture
   }
 
   private val basicAuthentication = authenticationHeader(username, password)
+  private val basicAuthenticationAndZipContentType = basicAuthentication  :+ "Content-Type" -> "application/zip"
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala]:
```suggestion
  private val basicAuthenticationAndZipContentType = "Content-Type" -> "application/zip" :: basicAuthentication
```

File: src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala [hunk 9]
```diff
@@ -546,7 +546,7 @@ class StoresServletSpec extends TestSupportFixture
     createZipWithInvalidOrEmptyRefBag(content)
     val uuid = "11111111-1121-1111-1111-111111111111"
     val bagId = BagId(UUID.fromString(uuid))
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala]: this line is never used

**Review by rvanheest — APPROVED** [review]
looks good! Here are some final changes in the tests.

File: docs/api/api.yml [hunk 10]
```diff
@@ -216,6 +216,8 @@ paths:
             * The bag is not virtually-valid.
             * The bag contained a `refbags.txt` file but it contained no data, or malformed data or
               references to non-existent bags.
+        415:
+          description: The request did not contain the appropiate Content-Type header ('application/zip')
```
Comment (janvanmansum) [inline docs/api/api.yml]:
```suggestion
          description: Unsupported media type
```

File: src/main/scala/nl.knaw.dans.easy.bagstore/package.scala [hunk 11]
```diff
@@ -50,6 +50,7 @@ package object bagstore {
   case class NoBagException(cause: Throwable) extends Exception("The provided input did not contain a bag", cause)
   case class InvalidBagException(bagId: BagId, msg: String) extends Exception(s"Bag $bagId is not a valid bag: $msg")
   case class NoRegularFileException(itemId: ItemId) extends Exception(s"Item $itemId is not a regular file.")
+  case class UnsupportedMediaTypeException(givenType: String, acceptedType: String) extends Exception(s"media type $givenType is not supported by this api. Supported types are '$acceptedType'")
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.bagstore/package.scala]:
```suggestion
  case class UnsupportedMediaTypeException(givenType: String, acceptedType: String) extends Exception(s"media type $givenType is not supported by this API. Supported types are '$acceptedType'")
```

**Review by janvanmansum — CHANGES_REQUESTED** [review]
Mostly fine. Have a look at the suggested changes.

File: src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala [hunk 12]
```diff
@@ -545,8 +545,7 @@ class StoresServletSpec extends TestSupportFixture
     val content = "invalid content"
     createZipWithInvalidOrEmptyRefBag(content)
     val uuid = "11111111-1121-1111-1111-111111111111"
-    val bagId = BagId(UUID.fromString(uuid))
-    put(s"/store1/bags/$uuid", body = Files.readAllBytes(testBagUnprunedEmptyRefBag), basicAuthentication) {
+    put(s"/store1/bags/$uuid", body = Files.readAllBytes(testBagUnprunedEmptyRefBag),headers =  basicAuthenticationAndZipContentType) {
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.bagstore/server/StoresServletSpec.scala]:
```suggestion
    put(s"/store1/bags/$uuid", body = Files.readAllBytes(testBagUnprunedEmptyRefBag), headers = basicAuthenticationAndZipContentType) {
```

**Review by janvanmansum — APPROVED** [review]

**Pull request merged**